src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,583 @@
+                                        });
+
+                                        await writer.FlushAsync().ConfigureAwait(false);
+                                        throw new InvalidOperationException($"Can't connect to shard '{db}' because: {reply.Message}");
+                                }
+
+                                return reply.Version;
+                            }
+                        },
+                        DestinationUrl = chosenUrl
+                    });
+
+                    if (supportedFeatures.ProtocolVersion <= 0)
+                        throw new InvalidOperationException($"{Options.SubscriptionName}: TCP negotiation resulted with an invalid protocol version:{supportedFeatures.ProtocolVersion}");
+
+
+                    _workers.Add(new SubscriptionServerWorker(stream, tcpClient, _serverStore));
+                }
+            }
+
+            await TryConnectSubscription();
+            _pendingConnectionScope.Dispose();
+
+            try
+            {
+                _activeConnectionScope = _connectionScope.For(SubscriptionOperationScope.ConnectionActive);
+
+                // refresh subscription data (change vector may have been updated, because in the meanwhile, another subscription could have just completed a batch)
+                SubscriptionState = await TcpConnection.ShardedContext.ShardedSubscriptionStorage.AssertSubscriptionConnectionDetails(SubscriptionId, _options.SubscriptionName);
+                foreach (var worker in _workers)
+                {
+                    using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+                    {
+                        using (var optionsJson = DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(Options, context))
+                        {
+                            await optionsJson.WriteJsonToAsync(worker.Stream, CancellationTokenSource.Token).ConfigureAwait(false);
+                            await worker.Stream.FlushAsync(CancellationTokenSource.Token).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+            catch
+            {
+                DisposeOnDisconnect.Dispose();
+                throw;
+            }
+
+            _connectionState = TcpConnection.ShardedContext.ShardedSubscriptionStorage.OpenSubscription(this);
+            _connectionState.PendingConnections.Add(this);
+
+            // Assert connection status
+            foreach (var worker in _workers)
+            {
+                using (_serverStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+                using (context.GetMemoryBuffer(out var buffer))
+                {
+                    var receivedMessage = await ReadNextObject(context, worker, buffer);
```

Comment:
And what if there isn't anything there yet?
You are not multiplexing between them